Creates a working CMake file for twbt as a `dfhack/plugins/external/` type plugin

cppcooper opened the pull request.

cppcooper committed:
Creates working cmake - requires modification like the make file

cppcooper commented:
No idea what to do about that build check failing. I guess it wants me to use a develop branch?? I figure you can test the cmake if you like the idea of having one that works and just bypass this check.

softmoth commented:
This is good. I *think* it could automatically read the dfhack version info from dfhack's CMakeLists.txt, so all you'd have to do is point it to the DFHack source dir (could even default that to `../dfhack`, probably) and it would just work. But that's a much bigger change, and this more conservative change is heading in the right direction, I think.

thurin commented:
I'll remove the build check.  dfhack guys were trying to set it up on their build system but it's obviously borked.  taking a look at the PR. thanks!

softmoth committed:
CMakeLists.txt: Rely on dfhack/CMake/Common.cmake

This uses CMake's INCLUDE() directive to pull in definitions from the
DFHack project, rather than duplicating them. It's intended to ensure
that the same compiler, essential compiler options, version numbers, and
so forth, are used between DFHack and TWBT. This makes TWBT build more
like a subproject of DFHack.

The LOAD_CACHE() operation is key to getting the configuration as the
user called it when building DFHack. Just doing INCLUDE() will source
the CMake rules, but does not copy any settings set when CMake generated
its build files. But LOAD_CACHE does that, so if needed we can get the
compiler used or other things. To start with, we load the installation
prefix.

cppcooper committed:
Uses CMAKE_INSTALL_PREFIX now instead of DF

cppcooper committed:
Merge branch 'cmake_common'

cppcooper committed:
Adds basic detection for dfhack directory based on commons setups

cppcooper committed:
Adds check for dfhack_SOURCE_DIR

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
+cmake_minimum_required(VERSION 3.22)
```
This is very new. Is there something you've done that actually requires this? For a couple points of reference (not that Ubuntu is the only relevant distro): Ubuntu 18.04 and 20.04 (both LTS) distribute CMake 3.10 and 3.18, respectively. DFHack claims to require 2.8.12 at a minimum, although I suspect the true minimum is a bit newer.

cppcooper replied:
No idea, I've just been using cmake since this version so I have no other baseline. edit: in the sense I've been writing/modifying cmake files

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
-SET( TWBT_VER "5.xx" )
+SET(CMAKE_VERBOSE_MAKEFILE ON)
+if(DEFINED dfhack_SOURCE_DIR)
+    set(DFHACK ${dfhack_SOURCE_DIR})
```
Do you need to support out-of-DFHack-tree builds with CMake? It seems to be adding a lot of complexity, which most other plugins avoid by living in the `dfhack/plugins` tree.

cppcooper replied:
:3 complexity? It's just a couple if statements

lethosor replied:
I'm referring to all of the logic in this file (not just this branch) that supports out-of-DFHack builds. It's a significant portion of the file.

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
+    set(DFHACK ../../../dfhack)
+else()
+    message("couldn't find a dfhack directory nearby, modify the cmake and put your dfhack path")
+    SET(DFHACK /your/path/dfhack)
```
If you're keeping this, I would suggest using a [cached variable](https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry) so that it can be edited by the user in CMakeCache.txt, which is not tracked in source control. DFHack has several examples of this.

The message probably ought to be a FATAL_ERROR as well, so that CMake doesn't proceed with a known-invalid path.

cppcooper replied:
good point

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
 
-SET( CMAKE_CXX_FLAGS "-m64 -O3 -DDFHACK_VERSION=\"${DFHACK_VERSION}\" -DDF_${DV_VS} -DTWBT_VER=\"\\\"${TWBT_VER}\\\"\"" )
+# include directories
+include_directories(${DFHACK}/library/include)
+include_directories(${DFHACK}/library/proto)
+include_directories(${DFHACK}/depends/protobuf)
+include_directories(${DFHACK}/depends/lua/include)
```
Redundant if df-twbt is inside dfhack/plugins

cppcooper replied:
I didn't want to complicate the conditional logic so I went for redundancy

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
-DFHACK_PLUGIN( twbt ${PROJECT_SRCS} )
+find_package(PNG)
+if(NOT PNG_FOUND)
+    # XXX Probably this is wrong. It appears df-twbt/libpng has precompiled
+    # libs for Windows? Probably it should be a submodule that has the
+    # actual libpng source (which builds with CMake).
```
My guess is that it was done deliberately because libpng is difficult to build, yet is stable enough that TWBT would almost never need to upgrade it. That's a valid decision to make - DFHack does the same thing for zlib and libruby. I personally wouldn't worry about it.

cppcooper replied:
I fully don't remember writing this. And it seems to build without it. maybe it's merged from that cmake_common

cppcooper replied:
indeed it is.

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
 
-# add_subdirectory( plugins ) Already included in dfhack?
+# twbt compile options
+add_compile_options("-Wno-error") # loads of warnings
+add_compile_options("-fpermissive") # taking address of rvalue
```
These will not work on Windows.

cppcooper replied:
good to know, it'll be a while til I can look up the corresponding options

cppcooper replied:
I've removed these.. so this branch on its own probably won't build now under dfhack.

lethosor commented on CMakeLists.txt:
```diff
@@ -1,22 +1,65 @@
+    #add_link_options("-L${DFHACK_BUILD}/library -ldfhack -ldfhack-version")
+    link_directories("${DFHACK_BUILD}/library")
+    add_library(${PLUGIN_TARGET} SHARED ${SRC} ${DEP})
+    target_link_libraries(${PLUGIN_TARGET} PUBLIC dfhack dfhack-version)
```
This is going to require the dfhack and dfhack-version targets to not only already be built, but to be built recently enough to be compatible with TWBT.

This is one of the bigger reasons why I think supporting only the in-DFHack-tree build approach is easier, and I imagine it's why other plugins only support it.

cppcooper replied:
if cmake knows the targets you're linking it builds them if they aren't available or up-to-date already, edit: though if that common file is never happening I'll try to obtain them from the cmake txt.. either through cmake or a cat/sed command

lethosor replied:
It can't build targets that aren't part of the current CMake project, which is the case in this branch.

cppcooper replied:
Fair, which then does require they be built before hand and present in the build directory...
Alright you know what? I yield.. I'll revise this to the bare bones situation, and update the build readme.. (correction, add one.. I'll add a build.md) I'll add the corresponding plugins sub-directory. `external` is that settled? or is `custom` preferred? I don't think you weighed in on that part

cppcooper replied:
This probably will require fixing the warnings though. And will probably make me sad when dfhack is updated for whatever the next version of DF will be, and the twbt offsets haven't been updated.. Everyone will have to move the directory or delete it else build all will fail for dfhack

lethosor replied:
I can handle the DFHack side of it. There are other solutions to temporarily disabling a plugin, like commenting out its entry in CMakeLists.txt (and the `external` one wouldn't be tracked anyway).

I'm fine with `external`.

cppcooper committed:
Simplifies cmake

cppcooper commented:
I've removed the compile options that bypassed warnings as errors. So this probably won't build.. but there is PR #5 which will make it build again.

cppcooper committed:
Adds BUILD.md

cppcooper committed:
Revises the build instructions

cppcooper commented:
- [ ] PR #5: fixes build errors since dfhack has the warnings as errors option on
- [ ] PR https://github.com/DFHack/dfhack/pull/2095: ensures the build instructions are precisely what's needed
- [ ] PR #2: the last piece of the puzzle

cppcooper committed:
Adds back the PNG package for missing symbols

cppcooper commented:
Ok, went back and tested the original. It works just fine with [#2095](https://github.com/DFHack/dfhack/pull/2095)

Strangely enough, after shuffling everything around this compiles but then doesn't link due to missing opengl symbols it looks like. A problem that didn't exist outside of the dfhack tree :raised_eyebrow:
